examples/pipeline_wav2letter/main.py

```diff
@@ -0,0 +1,668 @@
+
+    args = parser.parse_args()
+    logging.info(args)
+    return args
+
+
+def setup_distributed(rank, world_size):
+    os.environ["MASTER_ADDR"] = "localhost"
+    os.environ["MASTER_PORT"] = "12355"
+
+    # initialize the process group
+    torch.distributed.init_process_group("nccl", rank=rank, world_size=world_size)
+
+
+def model_length_function(tensor):
+    if tensor.shape[1] == 1:
+        # waveform mode
+        return int(tensor.shape[0]) // 160 // 2 + 1
+    return int(tensor.shape[0]) // 2 + 1
+
+
+def compute_error_rates(outputs, targets, decoder, language_model, metric):
+    output = outputs.transpose(0, 1).to("cpu")
+    output = decoder(output)
+
+    # Compute CER
+
+    output = language_model.decode(output.tolist())
+    target = language_model.decode(targets.tolist())
+
+    print_length = 20
+    for i in range(2):
+        # Print a few examples
+        output_print = output[i].ljust(print_length)[:print_length]
+        target_print = target[i].ljust(print_length)[:print_length]
+        logging.info("Target: %s    Output: %s", target_print, output_print)
+
+    cers = [levenshtein_distance(t, o) for t, o in zip(target, output)]
+    cers = sum(cers)
+    n = sum(len(t) for t in target)
+    metric["cer over target length"] = cers / n
+    metric["cumulative cer"] += cers
+    metric["total chars"] += n
+    metric["cumulative cer over target length"] = metric["cer"] / metric["total chars"]
+
+    # Compute WER
+
+    output = [o.split(language_model.char_space) for o in output]
+    target = [t.split(language_model.char_space) for t in target]
+
+    wers = [levenshtein_distance(t, o) for t, o in zip(target, output)]
+    wers = sum(wers)
+    n = sum(len(t) for t in target)
+    metric["wer over target length"] = wers / n
+    metric["cumulative wer"] += wers
+    metric["total words"] += n
+    metric["cumulative wer over target length"] = metric["wer"] / metric["total words"]
+
+
+def train_one_epoch(
+    model,
+    criterion,
+    optimizer,
+    scheduler,
+    data_loader,
+    decoder,
+    language_model,
+    device,
+    epoch,
+    clip_grad,
+    disable_logger=False,
+    reduce_lr_on_plateau=False,
+):
+
+    model.train()
+
+    metric = MetricLogger("train", disable=disable_logger)
+    metric["epoch"] = epoch
+
+    for inputs, targets, tensors_lengths, target_lengths in bg_iterator(
+        data_loader, maxsize=2
+    ):
+
+        start = time()
+        inputs = inputs.to(device, non_blocking=True)
+        targets = targets.to(device, non_blocking=True)
+
+        # keep batch first for data parallel
+        outputs = model(inputs).transpose(-1, -2).transpose(0, 1)
+
+        # CTC
+        # outputs: input length, batch size, number of classes (including blank)
+        # targets: batch size, max target length
+        # input_lengths: batch size
+        # target_lengths: batch size
+
+        loss = criterion(outputs, targets, tensors_lengths, target_lengths)
```

Comment:
Is this `loss` averaged over batch or summed?

Reply:
averaged by default -- but it's a good observation, i'll make it summed, and i'll average manually so that each iteration gets the same weight in case the last iteration of an epoch doesn't have the same number of samples